### ISSUE TITLE
Remove Rails/NotNullColumn cop

### ISSUE THREAD

**Issue by ixti** [bbatsov/rubocop#3963]
Please remove `Rails/NotNullColumn` cop. It makes no sense. See my comment: https://github.com/bbatsov/rubocop/pull/3415#issuecomment-272330214

Adding `null: false` adds `NOT NULL` constraint which is something most of us who care about DB integrity actually use. More than that, ActiveRecord have special error class that is raised if that constraint is violated: `ActiveRecord::NotNullViolation`.

Original issue that was used to introduce this cop was related to some specific database version and should not be used as a measure of sane defaults IMO.

**Comment by ixti** [comment]
After re-reading #3415 I figured out WHAT was happening. That issue is reproducible ONLY if you alter existing table with (!) existing data. In that case, OF COURSE you can't add `NOT NULL` constraint without providing default NOT NULL value, which is kinda expected. So again, I think that this cop does more harm than good.

**Comment by ixti** [comment]
Just to emphasize:

``` SQL
-- sqlite3 console
CREATE TABLE users ( name varchar NOT NULL );
INSERT INTO users ( name ) VALUES ( "ixti" );

-- attempt to insert row with NULL constraint violation fails:
INSERT INTO users ( name ) VALUES ( NULL );
-- Error: NOT NULL constraint failed: xxx.name

-- altering table with new column with not null constraint and no default value - fails
ALTER TABLE xxx ADD email  varchar NOT NULL;
-- Error: Cannot add a NOT NULL column with default value NULL
```

**Comment by ixti** [comment]
In any case, there are ways to add new column to database with not null constraint:

1. Add column with NULL
2. Fill in values for that field on all rows
3. alter column and set not null constraint

**Comment by mikegee** [comment]
Your argument is convincing.

I wonder why there haven't been more reports. The cop has been around since September (0.43.0).

**Comment by rrosenblum** [comment]
> I wonder why there haven't been more reports. The cop has been around since September (0.43.0).

My assumption is that it is because this is a new cop that probably has a large number of offenses in an existing project and therefore is disabled.

**Comment by pocke** [comment]
I think removing is not necessary.



```ruby
class CreateUsers < ActiveRecord::Migration[5.0]
  def change
    create_table :users do |t|
      t.string :name, null: false

      t.timestamps
    end

    User.create!(name: 'pocke')

    # This cop warns the following code.
    add_column :users, :email, :string, null: false
  end
end
```

In sqlite3, the above migration is failed with `ActiveRecord::StatementInvalid`. ( I think postgresql has same behavior.)
We can find the failing faster than execution the migration by the cop.


In MySQL, the above migration is not failed. The behavior is confusing.
We can notice the confusing behavior by the cop when we use MySQL.
The feature is useful if we support multiple RDBMS in one application(e.g. Gitlab).

**Comment by ixti** [comment]
@pocke Oh, indeed. Somehow I was sure it was failing for me on `create_table { |t| t.string :x, :null => false }`. But if it fails only upon attempt to add column with not null constraint on existing table - then it's fine, and I agree it's good to have it. :D

**Comment by ixti** [comment]
@pocke you're absolutely right that trying to add column to existing table with not null constraint fails on sqlite and postgresql (mysql makes me ¯\(°_o)/¯).

**Comment by ixti** [comment]
In fact PostgreSQL also allows you to add new column to existing table with NOT NULL constraint, but only if tables has no rows yet. If table has rows - adding a new field with not null constraint will fail.
